framework/src/main/java/org/checkerframework/framework/type/AnnotatedTypeFactory.java

```diff
@@ -3078,59 +3107,93 @@ public Set<AnnotationMirror> getWidenedAnnotations(
-     * @return the types of the two arguments
+     * @return the types of the two arguments (note that the underlying java types of both operands
+     *     will be converted to string if the binary operation is string concatenation)
      */
     public Pair<AnnotatedTypeMirror, AnnotatedTypeMirror> binaryTreeArgTypes(BinaryTree node) {
         return binaryTreeArgTypes(
-                getAnnotatedType(node.getLeftOperand()), getAnnotatedType(node.getRightOperand()));
+                TreeUtils.typeOf(node),
+                getAnnotatedType(node.getLeftOperand()),
+                getAnnotatedType(node.getRightOperand()));
     }
 
     /**
-     * Returns the types of the two arguments to the CompoundAssignmentTree, accounting for widening
-     * and unboxing if applicable.
+     * Returns the types of the two arguments to the CompoundAssignmentTree. Please refer to {@link
+     * #binaryTreeArgTypes(TypeMirror, AnnotatedTypeMirror, AnnotatedTypeMirror)} ) for more
+     * details.
      *
      * @param node a compound assignment tree
-     * @return the types of the two arguments
+     * @return the types of the two arguments (note that the underlying java types of the right
+     *     operand will be converted to string if the binary operation is string concatenation)
      */
     public Pair<AnnotatedTypeMirror, AnnotatedTypeMirror> compoundAssignmentTreeArgTypes(
             CompoundAssignmentTree node) {
         return binaryTreeArgTypes(
-                getAnnotatedType(node.getVariable()), getAnnotatedType(node.getExpression()));
+                TreeUtils.typeOf(node.getVariable()),
+                getAnnotatedType(node.getVariable()),
+                getAnnotatedType(node.getExpression()));
     }
 
     /**
-     * Returns the types of the two arguments to a binary operation, accounting for widening and
-     * unboxing if applicable.
+     * Returns the types of the two arguments to a binary operation. There are two special cases: 1.
+     * If both operands have numeric type, widening and unboxing will be applied accordingly. 2. If
+     * the binary operation is string concatenation (i.e., result is a string) and there is one
+     * operand of non-string type, we are going to return the type of its converted string by
+     * applying the declaration bounds of string type. Please check {@link
+     * #getAnnotationOrTypeDeclarationBound} for more details.
      *
+     * @param resultType the type of the result of a binary operation
      * @param left the type of the left argument of a binary operation
      * @param right the type of the right argument of a binary operation
-     * @return the types of the two arguments
+     * @return the types of the two arguments (note that the underlying java types of both operands
+     *     will be converted to string if the binary operation is string concatenation)
      */
     public Pair<AnnotatedTypeMirror, AnnotatedTypeMirror> binaryTreeArgTypes(
```

Comment:
The 2-argument `binaryTreeArgTypes` method is now no longer available. As it is a public method, we should be careful.
Can you add a deprecated version of the 2-arg method and explain how to transition?
Should this method be public at all?

Reply:
Yes, I can add the deprecated version back, and write some documentation for transition (i.e., use the wrapper functions instead). I think we should make the 3-arg method private.

Looing at the entire Github, I can't find any public external usage: https://github.com/search?q=binaryTreeArgTypes&type=code

Reply:
Done